### PR TITLE
[WIP] Github Gist backend

### PR DESCRIPTION
**Pull request by joostdevries**
**Do not merge**

Some initial work on using the Github API for this

- [x] Implement Github Oauth using Torii
- [x] Persist session in cookie
- [x] Fetch gist
- [x] Clean up Github code
- [ ] Write serializer/adapter for gist <---> file
- [ ] UI tweaks

### New pointers
- [ ] Block files whose content > 1 MB (will require additional API logic + it's insane to fiddle files that large)

### CODE REVIEW

File: app/services/github.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  _token: null,
+
+  setToken (token) {
+    this.set('_token', token);
+  },
+
+	apiCall (url, token) {
+		var token  = token || this.get('_token');
+
+    return new Ember.RSVP.Promise(function(resolve, reject){
```
Comment (alexspeller) [inline app/services/github.js]: Use ic.ajax instead here

File: app/torii-adapters/application.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+		});
+	},
+
+  open (authorization) {
+  	var cookie = this.get('cookie'),
+        github = this.get('github');
+
+    return new Ember.RSVP.Promise(function(resolve, reject){
+      Ember.$.ajax({
+        url: 'http://localhost:9999/authenticate/' + authorization.authorizationCode,
```
Comment (alexspeller) [inline app/torii-adapters/application.js]: Use ENV config for auth url

Comment (joostdevries) [inline app/torii-adapters/application.js]: yup that's on my list :-)

File: app/services/github.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+export default Ember.Object.extend({
+  _token: null,
+
+  setToken (token) {
+    this.set('_token', token);
+  },
+
+	apiCall (url, token) {
```
Comment (alexspeller) [inline app/services/github.js]: Use two spaces not hard tabs throughout, to avoid weirdness like this :)

File: app/torii-adapters/application.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+export default Ember.Object.extend({
+	github: Ember.inject.service('github'),
+
+	fetch () {
+		var cookie = this.get('cookie');
```
Comment (alexspeller) [inline app/torii-adapters/application.js]: Maybe we should use localStorage not cookie here? Not sure about this

File: app/gist/edit/route.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export default Em.Route.extend({
+	github: Ember.inject.service('github'),
+
+	model: function(params) {
+		return this.get('github').apiCall('/gists/' + params.id);
```
Comment (alexspeller) [inline app/gist/edit/route.js]: `get` is not necessary for injected services

File: app/gist/route.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+export default Em.Route.extend({
+  github: Ember.inject.service('github'),
+
+  beforeModel () {
+    var github = this.get('github'),
+        session = this.get('session');
```
Comment (alexspeller) [inline app/gist/route.js]: get is not necessary for injected services

File: package.json [hunk 7]
```diff
@@ -21,13 +21,15 @@
   "devDependencies": {
     "babel": "^4.6.6",
     "broccoli-asset-rev": "^2.0.0",
+    "broccoli-sass": "git://github.com/ef4/broccoli-sass",
```
Comment (alexspeller) [inline package.json]: I don't think this is necessary as ember-cli-bootstrap-sassy already pulls it in

Comment (joostdevries) [inline package.json]: don't remember adding it. weird.

File: app/gist/route.js (hunk 6 above)
Comment (joostdevries) [inline app/gist/route.js]: would expect that too but it doesn't work right now if I remove it.

File: app/torii-adapters/application.js (hunk 4 above)
Comment (joostdevries) [inline app/torii-adapters/application.js]: neither am I ;-).

**Comment by joostdevries** [comment]
@alexspeller cleaned up the code.

**Comment by alexspeller** [comment]
This is looking great, do you want me to merge now or wait for the other checklist items?

**Comment by joostdevries** [comment]
I'd rather have to complete flow working first. I might run into stuff forcing me to refactor.

File: app/lib/gist.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+  /**
+   * Finds a gist by id
+   * @return Promise
+   */
+  find (id) {
+    var self = this;
+
+    return this.github.find('/gists/' + id).then(function (json) {
+      return self.deserialize(json);
```
Comment (alexspeller) [inline app/lib/gist.js]: Use arrows instead of self = this

File: app/lib/gist.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    });
+  },
+
+  /**
+   * Saves the current gist
+   * @return Promise
+   */
+  save () {
+    var self = this;
```
Comment (alexspeller) [inline app/lib/gist.js]: Use arrows instead of self = this

**Comment by joostdevries** [comment]
@alexspeller maybe we should setup Travis for this?

**Comment by joostdevries** [comment]
@alexspeller pull && http://localhost:4200/f2c90713bdfdf54a262b

**Comment by joostdevries** [comment]
The demo app is not rendering yet and module resolving is not working atm.

**Comment by joostdevries** [comment]
@alexspeller I've created a `gists` branch. Maybe it's better to merge my work into that so we can both work on it?

**Comment by joostdevries** [comment]
@alexspeller two questions:
- It looks like the resolver is not always finding the right files. http://localhost:4200/202e3ab3483386daacd2 works but http://localhost:4200/f2c90713bdfdf54a262b doesn't. 
- Should we support css with this?

**Pull request merged**